txnkv/transaction/pipelined_flush.go

```diff
@@ -0,0 +1,448 @@
+	c.mu.RUnlock()
+	logutil.BgLogger().Info("[pipelined dml] transaction is committed")
+
+	if _, err := util.EvalFailpoint("pipelinedSkipResolveLock"); err == nil {
+		return nil
+	}
+
+	// async resolve the rest locks.
+	commitBo := retry.NewBackofferWithVars(c.store.Ctx(), CommitSecondaryMaxBackoff, c.txn.vars)
+	go c.resolveFlushedLocks(commitBo, c.pipelinedCommitInfo.pipelinedStart, c.pipelinedCommitInfo.pipelinedEnd)
+	return nil
+}
+
+// resolveRegionFlushedLocks resolves all locks for the given region.
+// If the region cache is stale,it reloads the region info and resolve the rest ranges.
+// The function returns the resolved regions.
+func (c *twoPhaseCommitter) resolveRegionFlushedLocks(bo *retry.Backoffer, region locate.RegionVerID, start, end []byte) (int, error) {
+	lreq := &kvrpcpb.ResolveLockRequest{
+		StartVersion:  c.startTS,
+		CommitVersion: atomic.LoadUint64(&c.commitTS),
+	}
+	req := tikvrpc.NewRequest(tikvrpc.CmdResolveLock, lreq, kvrpcpb.Context{
+		RequestSource: c.txn.GetRequestSource(),
+	})
+	req.RequestSource = c.txn.GetRequestSource()
+	for {
+		resp, err := c.store.SendReq(bo, req, region, client.ReadTimeoutShort)
+		if err != nil {
+			err = bo.Backoff(retry.BoRegionMiss, err)
+			if err != nil {
+				logutil.Logger(bo.GetCtx()).Error("send resolve lock request error", zap.Error(err))
+				return 0, err
+			}
+			continue
+		}
+		regionErr, err := resp.GetRegionError()
+		if err != nil {
+			logutil.Logger(bo.GetCtx()).Error("get region error failed", zap.Error(err))
+			return 0, err
+		}
+		if regionErr != nil {
+			err = bo.Backoff(retry.BoRegionMiss, errors.New(regionErr.String()))
+			if err != nil {
+				logutil.Logger(bo.GetCtx()).Error("send resolve lock get region error", zap.Error(err))
+				return 0, err
+			}
+			// reload the region
+			subRegions, err := c.store.GetRegionCache().LoadRegionsInKeyRange(bo, start, end)
+			if err != nil {
+				return 0, err
+			}
+			resolvedRegions := 0
+			for i, subRegion := range subRegions {
+				regionStart, regionEnd := subRegion.StartKey(), subRegion.EndKey()
+				if i == 0 {
+					regionStart = start
+				}
+				if i == len(subRegions)-1 {
+					regionEnd = end
+				}
+				resolved, err := c.resolveRegionFlushedLocks(bo, subRegion.VerID(), regionStart, regionEnd)
+				if err != nil {
+					return 0, err
+				}
+				resolvedRegions += resolved
+			}
+			return resolvedRegions, nil
+		}
+		if resp.Resp == nil {
+			logutil.Logger(bo.GetCtx()).Error("send resolve lock response body missing", zap.Error(errors.WithStack(tikverr.ErrBodyMissing)))
+			return 0, err
+		}
+		cmdResp := resp.Resp.(*kvrpcpb.ResolveLockResponse)
+		if keyErr := cmdResp.GetError(); keyErr != nil {
+			err = errors.Errorf("unexpected resolve err: %s", keyErr)
+			logutil.BgLogger().Error("resolveLock error", zap.Error(err))
+			return 0, err
+		}
+		return 1, nil
+	}
+}
+
+func (c *twoPhaseCommitter) resolveFlushedLocks(bo *retry.Backoffer, start, end []byte) {
+	const RESOLVE_CONCURRENCY = 8
```

Comment:
The `BatchExecutor` already has a token limit to control concurrency, is it possible to re-use it?

Reply:
Current `batchExecutor` is mutation based, since here we do not have any mutations, it's not applicable now.

I'll add a TODO which unify them by replace the mutation batch executor with task mutation batch executor.

Reply:
Agree.

Reply:
I think the range task runner is more suitable, it's used by GC resolve lock.

https://github.com/tikv/client-go/blob/0f8c594090485696741c84a1441cdacb2d14d597/txnkv/rangetask/range_task.go#L60-L73

Reply:
Now commit by range task runner.